File: src/agent_os_cli/lib/agent_os_cli/commands/deploy.ex
defmodule AgentOS.CLI.Commands.Deploy do
  @moduledoc """
  Deployment commands: docker, fly.

  These commands manage deploying agent-os itself (not individual agents)
  to Docker or Fly.io.
  """

  alias AgentOS.CLI.Output

  @doc "Routes deploy subcommands to their handlers."
  def run(["docker"], opts) do
    deploy_docker(opts)
  end

  def run(["fly" | rest], opts) do
    deploy_fly(rest, opts)
  end

  def run(_, _opts) do
    Output.info("""
    Usage: agent-os deploy <target> [options]

    Targets:
      docker                       Build and run with Docker Compose
      fly [--region <region>] [--app <name>]   Deploy to Fly.io
    """)
  end

  # --- Docker deployment ---

  defp deploy_docker(_opts) do
    Output.info("Building Docker image...")

    case System.cmd("docker", ["build", "-t", "agent-os", "."], stderr_to_stdout: true) do
      {output, 0} ->
        Output.info(output)
        Output.info("Starting containers with docker-compose...")

        case System.cmd("docker-compose", ["up", "-d"], stderr_to_stdout: true) do
          {up_output, 0} ->
            Output.info(up_output)
            Output.success("Docker deployment complete. Checking health...")
            check_local_health()

          {up_output, code} ->
            Output.error("docker-compose up failed (exit #{code}):\n#{up_output}")
        end

      {output, code} ->
        Output.error("docker build failed (exit #{code}):\n#{output}")
    end
  end

  # --- Fly.io deployment ---

  defp deploy_fly(args, _opts) do
    fly_opts = parse_fly_opts(args)
    region = Map.get(fly_opts, :region, "iad")
    app = Map.get(fly_opts, :app, "agent-os")

    Output.info("Deploying to Fly.io (app: #{app}, region: #{region})...")

    fly_args = ["deploy", "--region", region, "--app", app]

    case System.cmd("fly", fly_args, stderr_to_stdout: true) do
      {output, 0} ->
        Output.info(output)
        Output.success("Fly.io deployment complete. Checking health...")
        check_fly_health(app)

      {output, code} ->
        Output.error("fly deploy failed (exit #{code}):\n#{output}")
    end
  end

  # --- Health checks ---

  defp check_local_health do
    health_opts = %{host: "http://localhost:4000", api_key: nil, json: false}

    case AgentOS.CLI.HttpClient.get("/api/v1/health", health_opts) do
      {:ok, _} -> Output.success("Local health check passed")
      {:error, reason} -> Output.error("Local health check failed: #{inspect(reason)}")
    end
  end

  defp check_fly_health(app) do
    health_opts = %{host: "https://#{app}.fly.dev", api_key: nil, json: false}

    case AgentOS.CLI.HttpClient.get("/api/v1/health", health_opts) do
      {:ok, _} -> Output.success("Fly.io health check passed (#{app}.fly.dev)")
      {:error, reason} -> Output.error("Fly.io health check failed: #{inspect(reason)}")
    end
  end

  # --- Option parser ---

  defp parse_fly_opts(args), do: parse_kv(args, %{})

  defp parse_kv([], acc), do: acc
  defp parse_kv(["--region", v | rest], acc), do: parse_kv(rest, Map.put(acc, :region, v))
  defp parse_kv(["--app", v | rest], acc), do: parse_kv(rest, Map.put(acc, :app, v))
  defp parse_kv([_ | rest], acc), do: parse_kv(rest, acc)
end

File: cli/src/commands/deploy.js
/**
 * Deploy commands — docker, fly.
 */

import { execFileSync } from 'node:child_process';
import { resolve, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import * as http from '../http.js';
import * as out from '../output.js';

/** Resolve the repo root (cli/ is one level deep in agent-os). */
function repoRoot() {
  const __dirname = dirname(fileURLToPath(import.meta.url));
  return resolve(__dirname, '..', '..');
}

/**
 * Run a command safely using execFileSync (no shell injection).
 * @param {string} cmd - Command name
 * @param {string[]} cmdArgs - Command arguments
 * @param {string} cwd - Working directory
 */
function run(cmd, cmdArgs, cwd) {
  out.info(`> ${cmd} ${cmdArgs.join(' ')}`);
  execFileSync(cmd, cmdArgs, { cwd, stdio: 'inherit' });
}

/**
 * Deploy with Docker.
 * Builds the image and starts containers, then checks health.
 * @param {string[]} args - Positional arguments
 * @param {object} opts - Global CLI options
 */
export async function docker(args, opts) {
  const root = repoRoot();

  out.info('Building Docker image...');
  run('docker', ['build', '-t', 'agent-os', '.'], root);

  out.info('Starting containers...');
  run('docker-compose', ['up', '-d'], root);

  out.info('Checking health...');
  try {
    const result = await http.get('/api/v1/health', opts);
    out.success('Agent-OS is running (Docker).');
    out.json(result);
  } catch (e) {
    out.error(`Health check failed: ${e.message}`);
    out.info('Containers may still be starting. Try: agent-os health');
    process.exit(1);
  }
}

/**
 * Deploy to Fly.io.
 * Runs fly deploy with optional --region and --app flags, then checks health.
 * @param {string[]} args - Positional arguments
 * @param {object} opts - Global CLI options
 */
export async function fly(args, opts) {
  const root = repoRoot();
  const region = opts.region || 'iad';
  const app = opts.app || 'agent-os';

  out.info(`Deploying to Fly.io (app=${app}, region=${region})...`);
  run('fly', ['deploy', '--region', region, '--app', app], root);

  const flyHost = `https://${app}.fly.dev`;
  out.info(`Checking health at ${flyHost}...`);
  try {
    const result = await http.get('/api/v1/health', { ...opts, host: flyHost });
    out.success(`Agent-OS is running on Fly.io (${flyHost}).`);
    out.json(result);
  } catch (e) {
    out.error(`Health check failed: ${e.message}`);
    out.info('Deployment may still be in progress. Try: agent-os health');
    process.exit(1);
  }
}
